crates/sui-core/src/authority/transaction_deferral.rs

```diff
@@ -0,0 +1,199 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use narwhal_types::Round;
+use serde::{Deserialize, Serialize};
+use sui_types::base_types::ObjectID;
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, PartialOrd, Ord, Serialize, Deserialize)]
+pub enum DeferralKey {
+    // For transactions deferred until randomness DKG has completed.
+    RandomnessDkg {
+        deferred_from_round: Round,
+    },
+    // ConsensusRound deferral key requires both the round to which the tx should be deferred (so that
+    // we can efficiently load all txns that are now ready), and the round from which it has been
+    // deferred (so that multiple rounds can efficiently defer to the same future round).
+    ConsensusRound {
+        future_round: Round,
+        deferred_from_round: Round,
+    },
+}
+
+impl DeferralKey {
+    pub fn new_for_randomness(deferred_from_round: Round) -> Self {
+        Self::RandomnessDkg {
+            deferred_from_round,
+        }
+    }
+
+    pub fn new_for_consensus_round(future_round: Round, deferred_from_round: Round) -> Self {
+        Self::ConsensusRound {
+            future_round,
+            deferred_from_round,
+        }
+    }
+
+    pub fn full_range_for_randomness() -> (Self, Self) {
+        (
+            Self::RandomnessDkg {
+                deferred_from_round: 0,
+            },
+            Self::RandomnessDkg {
+                deferred_from_round: u64::MAX,
+            },
+        )
+    }
+
+    // Returns a range of deferral keys that are deferred up to the given consensus round.
+    pub fn range_for_up_to_consensus_round(consensus_round: Round) -> (Self, Self) {
+        (
+            Self::ConsensusRound {
+                future_round: 0,
+                deferred_from_round: 0,
+            },
+            Self::ConsensusRound {
+                future_round: consensus_round.checked_add(1).unwrap(),
+                deferred_from_round: 0,
+            },
+        )
+    }
+}
+
+#[derive(Debug)]
+pub enum DeferralReason {
+    RandomnessNotReady,
+
+    // The list of objects are congested objects.
+    SharedObjectCongestion(Vec<ObjectID>),
+}
+
+pub fn transaction_deferral_within_limit(
```

Comment:
To save some time to review: only this function is newly created. Everything else in this file was moved from authority_per_epoch_store.rs